### ISSUE TITLE
many-to-many join via intermediate object

### ISSUE THREAD

**Issue by bradzacher** [join-monster/join-monster#122]
I have two tables: users, and tags.
There exists a many-to-many relationship between them.
The junction table is called user_tags.

When a tag is added to a user, it can be added as a tag in one of three categories: 'a', 'b', or 'c'.
What I want is an object structure so that it separates the tags based on the category they were added to:
```gql
query {
  Users {
    id
    tags {
      a_tags {
        id
      }
      b_tags {
        id
      }
      c_tags {
        id
      }
    }
  }
}
```

The table structures are:

 users | user_tags | tags 
 --- | --- | --- 
 id | id | id 
 ... | user_id | ... 
 ... | tag_id | ... 
 ... | type | ... 

what i've got is three graphql object schemas;
```JS
const Tags = new GraphQLObjectType({
  sqlTable: 'tags',
  name: 'Tags',
  uniqueKey: 'id',
  fields: {
    id: {
      type: GraphQLInt
    }
  }
})
const UserTags = new GraphQLObjectType({
  sqlTable: 'user_tags',
  name: 'UserTags',
  uniqueKey: 'id',
  fields: {
    a_tags: {
      type: new GraphQLList(Tags),
      sqlJoin: (ut, t) => `${ut}.tag_id = ${t}.id and ${ut}.type = 'a'`
    },
    b_tags: {
      type: new GraphQLList(Tags),
      sqlJoin: (ut, t) => `${ut}.tag_id = ${t}.id and ${ut}.type = 'b'`
    },
    c_tags: {
      type: new GraphQLList(Tags),
      sqlJoin: (ut, t) => `${ut}.tag_id = ${t}.id and ${ut}.type = 'c'`
    },
  }
})
const Users = new GraphQLObjectType({
  sqlTable: 'users',
  name: 'Users',
  uniqueKey: 'id',
  fields: {
    id: {
      type: GraphQLInt
    },
    tags: {
      type: UserTags,
      sqlJoin: (u, ut) => `${u}.id = ${ut}.user_id`
    }
  }
})
```

when I run a query through, join-monster fails to hydrate if there are some tags for any type missing for any user.

```
[SERVER] TypeError: Cannot read property 'a_tags' of undefined
[SERVER]     at _nest (<redacted>/node_modules/nesthydrationjs/NestHydrationJS.js:176:14)
[SERVER]     at NestHydrationJS.nest (<redacted>/node_modules/nesthydrationjs/NestHydrationJS.js:202:4)
[SERVER]     at result.then.rows (<redacted>/node_modules/join-monster/dist/util.js:173:40)
[SERVER]     at process._tickCallback (internal/process/next_tick.js:109:7)
```

Looking at the join monster call, it happily builds this schema into a sql query:
```SQL
SELECT
  `Users`.`id` AS `id`,
  `user_tags`.`id` AS `user_tags__id`,
  `a_tags`.`id` AS `user_tags__a_tags__id`,
  `b_tags`.`id` AS `user_tags__b_tags__id`,
  `c_tags`.`id` AS `user_tags__c_tags__id`,
FROM users `Users`
LEFT JOIN user_tags `user_tags` ON `Users`.id = `user_tags`.user_id
LEFT JOIN tags `a_tags` ON `user_tags`.tag_id = `a_tags`.id AND `user_tags`.type = 'a'
LEFT JOIN tags `b_tags` ON `user_tags`.tag_id = `b_tags`.id AND `user_tags`.type = 'b'
LEFT JOIN tags `c_tags` ON `user_tags`.tag_id = `c_tags`.id AND `user_tags`.type = 'c'
```

it's just failing to hydrate.

**Comment by acarl005** [comment]
@bradzacher Can you post the version of `join-monster` you're using?

Also, can you double-check this part of your set-up?

```js
const Users = new GraphQLObjectType({
  sqlTable: 'users',
  name: 'Users',
  uniqueKey: 'id',
  fields: {
    tags: {
      // ↓↓↓ should this be a GraphQLList?
      type: UserTags,
      // ↑↑↑
      sqlJoin: (u, ut) => `${u}.id = ${ut}.user_id`
    }
  }
})

```

**Comment by bradzacher** [comment]
sorry, yeah - it was a typo. the code is just a truncated form of my real schema. i've fixed the comment.

I'm running version 1.2.1-beta.9

**Comment by acarl005** [comment]
Can you run your app with this environment variable: `DEBUG=join-monster`, trigger the bug again, and post the additional output?

**Comment by bradzacher** [comment]
okay, sorry, I tore down my config shortly after this, so I was wrong in the correction - the original code was what it is supposed to be - Users does not have a list of UserTags - it just has one.

I just tried setting it to list. It works, but it doesn't give me the result that I would like.
You end up with an object where `User.tags` is an array, and `User.tags[i].[a|b|c]_tags` are lists, each one having length one or zero.

my hope is that with the query provided, I can get a User back and `User.tags` is an object, `User.tags.[a|b|c]_tags` are all lists.


I've created a standalone, working gist of the example: https://gist.github.com/bradzacher/f58770b3564891700718a90a6b941601

It's configured to have the `User` `tags` attribute type to be list of `UserTags`. This works fine, but you can see the problem with the result:
```JSON
{
    "data": {
        "Users": [
            {
                "id": 1,
                "tags": [
                    {
                        "a_tags": [
                            {
                                "id": 3
                            }
                        ],
                        "b_tags": [],
                        "c_tags": []
                    },
                    {
                        "a_tags": [
                            {
                                "id": 4
                            }
                        ],
                        "b_tags": [],
                        "c_tags": []
                    },
                    {
                        "a_tags": [
                            {
                                "id": 5
                            }
                        ],
                        "b_tags": [],
                        "c_tags": []
                    },
                    {
                        "a_tags": [],
                        "b_tags": [
                            {
                                "id": 6
                            }
                        ],
                        "c_tags": []
                    },
                    {
                        "a_tags": [],
                        "b_tags": [],
                        "c_tags": [
                            {
                                "id": 7
                            }
                        ]
                    }
                ]
            },
            {
                "id": 2,
                "tags": []
            }
        ]
    }
}
```

If you change it to not be a list (schema.js, line 41), it will error out:
```
f58770b3564891700718a90a6b941601 bradzacher$ node index.js
  join-monster
  join-monster SQL_AST
  join-monster  { type: 'table',
  name: 'users',
  as: 'Users',
  children:
   [ { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
     { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
     { type: 'table',
       name: 'user_tags',
       as: 'tags',
       children:
        [ { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
          { type: 'table',
            name: 'tags',
            as: 'a_tags',
            children:
             [ { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
               { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
               { type: 'columnDeps', names: {} } ],
            fieldName: 'a_tags',
            grabMany: true,
            sqlJoin: [Function: sqlJoin] },
          { type: 'table',
            name: 'tags',
            as: 'b_tags',
            children:
             [ { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
               { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
               { type: 'columnDeps', names: {} } ],
            fieldName: 'b_tags',
            grabMany: true,
            sqlJoin: [Function: sqlJoin] },
          { type: 'table',
            name: 'tags',
            as: 'c_tags',
            children:
             [ { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
               { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
               { type: 'columnDeps', names: {} } ],
            fieldName: 'c_tags',
            grabMany: true,
            sqlJoin: [Function: sqlJoin] },
          { type: 'columnDeps', names: {} } ],
       fieldName: 'tags',
       grabMany: false,
       sqlJoin: [Function: sqlJoin] },
     { type: 'columnDeps', names: {} } ],
  fieldName: 'Users',
  grabMany: true } +0ms
  join-monster
  join-monster SQL
  join-monster  SELECT
  "Users"."id" AS "id",
  "tags"."id" AS "tags__id",
  "a_tags"."id" AS "tags__a_tags__id",
  "b_tags"."id" AS "tags__b_tags__id",
  "c_tags"."id" AS "tags__c_tags__id"
FROM users "Users"
LEFT JOIN user_tags "tags" ON "Users".id = "tags".user_id
LEFT JOIN tags "a_tags" ON "tags".tag_id = "a_tags".id and "tags".type = 'a'
LEFT JOIN tags "b_tags" ON "tags".tag_id = "b_tags".id and "tags".type = 'b'
LEFT JOIN tags "c_tags" ON "tags".tag_id = "c_tags".id and "tags".type = 'c' +6ms
  join-monster
  join-monster SHAPE_DEFINITION
  join-monster  [ { id: 'id',
    tags:
     { id: 'tags__id',
       a_tags: [ { id: 'tags__a_tags__id' } ],
       b_tags: [ { id: 'tags__b_tags__id' } ],
       c_tags: [ { id: 'tags__c_tags__id' } ] } } ] +3ms
  join-monster
  join-monster RAW DATA
  join-monster  [ { id: 1,
    tags__id: 1,
    tags__a_tags__id: 3,
    tags__b_tags__id: null,
    tags__c_tags__id: null },
  { id: 1,
    tags__id: 2,
    tags__a_tags__id: 4,
    tags__b_tags__id: null,
    tags__c_tags__id: null },
  { id: 1,
    tags__id: 3,
    tags__a_tags__id: 5,
    tags__b_tags__id: null,
    tags__c_tags__id: null },
  { id: 1,
    tags__id: 4,
    tags__a_tags__id: null,
    tags__b_tags__id: 6,
    tags__c_tags__id: null },
  { id: 1,
    tags__id: 5,
    tags__a_tags__id: null,
    tags__b_tags__id: null,
    tags__c_tags__id: 7 },
  { id: 2,
    tags__id: null,
    tags__a_tags__id: null,
    tags__b_tags__id: null,
    tags__c_tags__id: null } ] +4ms
  join-monster 6 rows... +0ms
```

**Comment by acarl005** [comment]
So in GraphQL, `UserTag` is a singular field on `User`. Is it possible that the `user_tags` table actually has multiple rows for a user, despite the intention in GraphQL for it to be singular? If the tables have multiple rows for a join that Join Monster expects to be singular, it can trigger this exception: CoursePark/NestHydrationJS#15. It's an issue I reported on one of Join Monster's dependencies. It's not exactly a bug. In the interest of memory efficiency, this case is unable to be handled.

**Comment by bradzacher** [comment]
ahk, that's all good then.

is there any way around this to do this sort of thing then?

For example, I often group fields together into a "fake" sub-object by using sqlDeps + a resolve function on a field to select the fields, and construct the object, respectively.

Is there a way to do this sort of thing for constructing "fake" sub-arrays as well?

I.e. Have the sql join to fetch all of the tags for a user as a single array, then use a resolve function to filter the array into three separate arrays within a subobject (thus achieving the same structure).

I could technically have separate arrays on the user object, but I would rather group them on a "sub-object".


it's possible to do if you change the UserTag s schema so that its table is User, and it has the sub-queries:

```JS
const Tags = new GraphQLObjectType({
  sqlTable: 'tags',
  name: 'Tags',
  uniqueKey: 'id',
  fields: {
    id: {
      type: GraphQLInt
    }
  }
})
const UserTags = new GraphQLObjectType({
  sqlTable: 'users',
  name: 'UserTags',
  uniqueKey: 'id',
  fields: {
    a_tags: {
      type: new GraphQLList(Tags),
      junctionTable: 'user_tags',
      sqlJoins: [
          (u, ut) => `${u}.id = ${ut}.user_id AND ${ut}.type = 'a'`,
          (ut, t) => `${ut}.tag_id = ${t}.id`,
      ],
    },
    b_tags: {
      type: new GraphQLList(Tags),
      junctionTable: 'user_tags',
      sqlJoins: [
          (u, ut) => `${u}.id = ${ut}.user_id AND ${ut}.type = 'a'`,
          (ut, t) => `${ut}.tag_id = ${t}.id`,
      ],
    },
    c_tags: {
      type: new GraphQLList(Tags),
      junctionTable: 'user_tags',
      sqlJoins: [
          (u, ut) => `${u}.id = ${ut}.user_id AND ${ut}.type = 'a'`,
          (ut, t) => `${ut}.tag_id = ${t}.id`,
      ],
    },
  }
})
const Users = new GraphQLObjectType({
  sqlTable: 'users',
  name: 'Users',
  uniqueKey: 'id',
  fields: {
    id: {
      type: GraphQLInt,
    },
    tags: {
      type: (UserTags),
      sqlJoin: (u1, u2) => `${u1}.id = ${u2}.id`,
    }
  }
})
```

but, of course, this generates redundant joins in the query:
```SQL
SELECT
  "Users"."id" AS "id",
  "tags"."id" AS "tags__id",
  "a_tags"."id" AS "tags__a_tags__id",
  "b_tags"."id" AS "tags__b_tags__id",
  "c_tags"."id" AS "tags__c_tags__id"
FROM users "Users"
LEFT JOIN users "tags" ON "Users".id = "tags".id
LEFT JOIN user_tags "user_tags" ON "tags".id = "user_tags".user_id AND "user_tags".type = 'a'
LEFT JOIN tags "a_tags" ON "user_tags".tag_id = "a_tags".id
LEFT JOIN user_tags "user_tags$" ON "tags".id = "user_tags$".user_id AND "user_tags$".type = 'a'
LEFT JOIN tags "b_tags" ON "user_tags$".tag_id = "b_tags".id
LEFT JOIN user_tags "user_tags$$" ON "tags".id = "user_tags$$".user_id AND "user_tags$$".type = 'a'
LEFT JOIN tags "c_tags" ON "user_tags$$".tag_id = "c_tags".id
```

**Comment by Reggino** [comment]
@bradzacher I ran into this issue too and i think i worked out a fix. Could you please try your code using my fork at https://github.com/reggino/join-monster ?

The reason why i don't have a PR here yet, is because i'm waiting for a PR to land here: 
https://github.com/CoursePark/NestHydrationJS/pull/16

and didn't have the time to write a proper test. Your issue may help me there!

**Comment by acarl005** [comment]
I forked the NestHydrationJS library. Newest version **v2.0.2** should now be able to handle this case.

**Comment by acarl005** [comment]
Assuming the fix worked.

**Comment by bradzacher** [comment]
sorry - I hadn't gotten around to revisiting this after your fix.

it doesn't seem to achieve the correct outcome.

providing the following query:
```graphql
query a {
  Users(id:432) {
    id
    access_tags {
      a {
        id
      }
      b {
        id
      }
      c {
        id
      }
    }
  }
}
```

Again, the SQL generated is:
```SQL
SELECT
  `Users`.`user_id` AS `user_id`,
  `Users`.`user_id` AS `id`,
  `access_tag`.`id` AS `access_tag__id`,
  `a`.`id` AS `access_tag__a__id`,
  `b`.`id` AS `access_tag__b__id`,
  `c`.`id` AS `access_tag__c__id`
FROM user `Users`
LEFT JOIN user_access `access_tag` ON `Users`.user_id = `access_tag`.user_id
LEFT JOIN terms `a` ON `access_tag`.term_id = `a`.id AND `access_tag`.object_name = 'a'
LEFT JOIN terms `b` ON `access_tag`.term_id = `b`.id AND `access_tag`.object_name = 'b'
LEFT JOIN terms `c` ON `access_tag`.term_id = `c`.id AND `access_tag`.object_name = 'c'
WHERE `Users`.user_id = 432
```

When run against the database you get the correct data back (note that there are no `c` type tags for this user..):

| user_id | id | access_tag__id | access_tag__a__id | access_tag__b__id | access_tag__c__id |
| ------- | -- | --------------- | ------------------ | ------------------ | ------------------ |
| 432 | 432 | 984 | 1 | NULL | NULL |
| 432 | 432 | 992 | NULL | 2 | NULL |
| 432 | 432 | 993 | NULL | 17 | NULL |
| 432 | 432 | 985 | 32 | NULL | NULL |
| 432 | 432 | 986 | 33 | NULL | NULL |
| 432 | 432 | 987 | 41 | NULL | NULL |
| 432 | 432 | 994 | NULL | 45 | NULL |
| 432 | 432 | 995 | NULL | 48 | NULL |
| 432 | 432 | 988 | 55 | NULL | NULL |
| 432 | 432 | 989 | 63 | NULL | NULL |
| 432 | 432 | 990 | 64 | NULL | NULL |
| 432 | 432 | 996 | NULL | 66 | NULL |
| 432 | 432 | 991 | 80 | NULL | NULL |
| 432 | 432 | 997 | NULL | 81 | NULL |
| 432 | 432 | 998 | NULL | 82 | NULL |
| 432 | 432 | 999 | NULL | 83 | NULL |
| 432 | 432 | 1000 | NULL | 106 | NULL |

However, the hydration produces the following wrong result:
```JSON
{
  "data": {
    "Users": [
      {
        "id": 432,
        "access_tags": {
          "a": [
            {
              "id": 1
            }
          ],
          "b": [],
          "c": []
        }
      }
    ]
  }
}
```

below is the debug output:
```
  join-monster
  join-monster SQL_AST
  join-monster  { args: { id: 432 },
  type: 'table',
  name: 'user',
  as: 'Users',
  children:
   [ { type: 'column',
       name: 'user_id',
       fieldName: 'user_id',
       as: 'user_id' },
     SQLASTNode {
       args: {},
       type: 'column',
       name: 'user_id',
       fieldName: 'id',
       as: 'id' },
     SQLASTNode {
       args: {},
       type: 'table',
       name: 'user_access',
       as: 'access_tag',
       children:
        [ { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
          SQLASTNode {
            args: {},
            type: 'table',
            name: 'tags',
            as: 'a',
            children:
             [ { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
               SQLASTNode { args: {}, type: 'column', name: 'id', fieldName: 'id', as: 'id' } ],
            fieldName: 'a',
            grabMany: true,
            sqlJoin: [Function: sqlJoin] },
          SQLASTNode {
            args: {},
            type: 'table',
            name: 'tags',
            as: 'b',
            children:
             [ { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
               SQLASTNode { args: {}, type: 'column', name: 'id', fieldName: 'id', as: 'id' } ],
            fieldName: 'b',
            grabMany: true,
            sqlJoin: [Function: sqlJoin] },
          SQLASTNode {
            args: {},
            type: 'table',
            name: 'tags',
            as: 'c',
            children:
             [ { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
               SQLASTNode { args: {}, type: 'column', name: 'id', fieldName: 'id', as: 'id' } ],
            fieldName: 'c',
            grabMany: true,
            sqlJoin: [Function: sqlJoin] } ],
       fieldName: 'access_tags',
       grabMany: false,
       sqlJoin: [Function] } ],
  fieldName: 'Users',
  grabMany: true,
  where: [Function: where] } +49s
  join-monster
  join-monster SQL
  join-monster  SELECT
  `Users`.`user_id` AS `user_id`,
  `Users`.`user_id` AS `id`,
  `access_tag`.`id` AS `access_tag__id`,
  `a`.`id` AS `access_tag__a__id`,
  `b`.`id` AS `access_tag__b__id`,
  `c`.`id` AS `access_tag__c__id`
FROM user `Users`
LEFT JOIN user_access `access_tag` ON `Users`.user_id = `access_tag`.user_id
LEFT JOIN tags `a` ON `access_tag`.tag_id = `a`.id AND `access_tag`.object_name = 'a'
LEFT JOIN tags `b` ON `access_tag`.tag_id = `b`.id AND `access_tag`.object_name = 'b'
LEFT JOIN tags `c` ON `access_tag`.tag_id = `c`.id AND `access_tag`.object_name = 'c'
WHERE `Users`.user_id = 432 +2ms
  join-monster
  join-monster SHAPE_DEFINITION
  join-monster  [ { user_id: 'user_id',
    id: 'id',
    access_tags:
     { id: 'access_tag__id',
       a: [ { id: 'access_tag__a__id' } ],
       b: [ { id: 'access_tag__b__id' } ],
       c: [ { id: 'access_tag__c__id' } ] } } ] +0ms
  join-monster
  join-monster RAW DATA
  join-monster  [ TextRow {
    user_id: 432,
    id: 432,
    access_tag__id: 984,
    access_tag__a__id: 1,
    access_tag__b__id: null,
    access_tag__c__id: null },
  TextRow {
    user_id: 432,
    id: 432,
    access_tag__id: 992,
    access_tag__a__id: null,
    access_tag__b__id: 2,
    access_tag__c__id: null },
  TextRow {
    user_id: 432,
    id: 432,
    access_tag__id: 993,
    access_tag__a__id: null,
    access_tag__b__id: 17,
    access_tag__c__id: null },
  TextRow {
    user_id: 432,
    id: 432,
    access_tag__id: 985,
    access_tag__a__id: 32,
    access_tag__b__id: null,
    access_tag__c__id: null },
  TextRow {
    user_id: 432,
    id: 432,
    access_tag__id: 986,
    access_tag__a__id: 33,
    access_tag__b__id: null,
    access_tag__c__id: null },
  TextRow {
    user_id: 432,
    id: 432,
    access_tag__id: 987,
    access_tag__a__id: 41,
    access_tag__b__id: null,
    access_tag__c__id: null },
  TextRow {
    user_id: 432,
    id: 432,
    access_tag__id: 994,
    access_tag__a__id: null,
    access_tag__b__id: 45,
    access_tag__c__id: null },
  TextRow {
    user_id: 432,
    id: 432,
    access_tag__id: 995,
    access_tag__a__id: null,
    access_tag__b__id: 48,
    access_tag__c__id: null } ] +5ms
  join-monster 17 rows... +0ms
  join-monster
  join-monster SHAPED_DATA
  join-monster  [ { user_id: 432,
    id: 432,
    access_tags: { id: 984, a: [ { id: 1 } ], b: [], c: [] } } ] +1ms
```

**Comment by Reggino** [comment]
@bradzacher Could you test your code using my fork at https://github.com/reggino/join-monster ? 

The PR (https://github.com/CoursePark/NestHydrationJS/pull/17) is still open, so i can't make a proper PR here yet, since the changes depend on that.

**Comment by acarl005** [comment]
@Reggino I've forked that library so now join-monster depends on [this fork](https://github.com/stems/nesthydrationjs). Perhaps you could patch it with all your changes?

EDIT: as an aside, I forked an older version of that library as our benchmarking noticed a drop in perf after v0.4.0.

**Comment by Reggino** [comment]
@acarl005 While reviewing i noticed my changes are merged in one way or another. 

My NestHydrationJS changes are applied here https://github.com/CoursePark/NestHydrationJS/commit/651887f68d68e00f188696c6ef181d043f78be24 . My join-monster changes are deprecated since this commit https://github.com/stems/join-monster/commit/512069b22090e1441e9ff29f51988695385ff566

I guess this is not going to solve the issue...

**Comment by dvalentiate** [comment]
Hey Reggino, sorry for all the trouble in not merging your PR much earlier. And thanks to you guys for tracking down, and fixing a bug in my library. published as 1.0.3.

**Comment by nicoabie** [comment]
This caught my attention and I did a couple of tests.
It is true that there was a bug in https://github.com/CoursePark/NestHydrationJS fixed by @Reggino and then merged by @dvalentiate which was related.

But @bradzacher code was inconsistent with his idea. because of how user, user_tags and tags table were created there is NOT only one user_tag per user. there are many and that is why join-monster correctly works when setting user tags to be list.

Nonetheless there is a way to make what he wanted to work and that is setting the uniqueKey of user tags to be user_id instead of id. 

I created a repro demonstrating that https://github.com/agiletiger/join-monster-test-issue-122

And here you have a screenshot as well.
<img width="1920" alt="image" src="https://github.com/join-monster/join-monster/assets/2797992/b92dfc1e-d286-4590-9c57-15ed244268f0">


So I believe we can close this issue as there is a way to make it work.
